src/mlpack/methods/reinforcement_learning/ppo_impl.hpp

```diff
@@ -0,0 +1,216 @@
+  ActorNetworkType,
+  CriticNetworkType,
+  UpdaterType,
+  PolicyType,
+  ReplayType
+>::PPO(TrainingConfig config,
+       ActorNetworkType actorNetwork,
+       CriticNetworkType criticNetwork,
+       PolicyType policy,
+       ReplayType replayMethod,
+       UpdaterType updater,
+       EnvironmentType environment):
+  config(std::move(config)),
+  actorNetwork(std::move(actorNetwork)),
+  criticNetwork(std::move(criticNetwork)),
+  updater(std::move(updater)),
+  policy(std::move(policy)),
+  replayMethod(std::move(replayMethod)),
+  environment(std::move(environment))
+{
+  if (actorNetwork.Parameters().is_empty())
+    actorNetwork.ResetParameters();
+  if (criticNetwork.Parameters().is_empty())
+    criticNetwork.ResetParameters();
+
+  this->updater.Initialize(actorNetwork.Parameters().n_rows,
+                           actorNetwork.Parameters().n_cols);
+  this->oldActorNetwork = actorNetwork;
+}
+
+template<
+  typename EnvironmentType,
+  typename ActorNetworkType,
+  typename CriticNetworkType,
+  typename UpdaterType,
+  typename PolicyType,
+  typename ReplayType
+>
+void PPO<
+  EnvironmentType,
+  ActorNetworkType,
+  CriticNetworkType,
+  UpdaterType,
+  PolicyType,
+  ReplayType
+>::Update()
+{
+  // Sample from previous experience.
+  arma::mat sampledStates;
+  std::vector<ActionType> sampledActions;
+  arma::colvec sampledRewards;
+  arma::mat sampledNextStates;
+  arma::icolvec isTerminal;
+  arma::colvec discountedRewards;
+
+  replayMethod.Sample(sampledStates, sampledActions, sampledRewards,
+      sampledNextStates, isTerminal);
+
+  arma::mat nextActionValues;
+  double values;
+  criticNetwork.Predict(sampledNextStates, nextActionValues);
+  for (size_t i = sampledRewards.n_cols; i > 0; --i)
+  {
+    values = sampledRewards[i-1] + values * config.Discount();
+    discountedRewards.insert_rows(discountedRewards.n_cols, values);
+  }
+
+  arma::mat actionValues, advantages, gradients;
+  criticNetwork.Predict(sampledStates, actionValues);
+
+  advantages = discountedRewards - actionValues;
+
+  criticNetwork.Backward(advantages, gradients);
+
+  updater.Update(criticNetwork.Parameters(), config.StepSize(), gradients);
+
+  // update the critic
+  criticNetwork.Backward(actionValues, sampledRewards);
+
+  // update the actor
+}
+
+template<
+  typename EnvironmentType,
+  typename ActorNetworkType,
+  typename CriticNetworkType,
+  typename UpdaterType,
+  typename PolicyType,
+  typename ReplayType
+>
+double PPO<
+  EnvironmentType,
+  ActorNetworkType,
+  CriticNetworkType,
+  UpdaterType,
+  PolicyType,
+  ReplayType
+>::Step()
+{
+  // Get the action value for each action at current state.
+  arma::mat actionValue;
+  actorNetwork.Predict(state.Encode(), actionValue);
+
+  arma::mat sigma, mu;
+  ann::TanhFunction::Fn(actionValue.col(0), sigma);
+  ann::SoftplusFunction::Fn(actionValue.col(1), mu);
+
+  distribution::GaussianDistribution normalDist =
+    distribution::GaussianDistribution(sigma, mu);
```

Comment:
Looks nice, just another minor style issue, this should be tabbed twice so 4 spaces.